libs/agno/agno/agent/agent.py

```diff
@@ -569,347 +569,426 @@ def has_team(self) -> bool:
-        7. Generate a response from the Model (includes running function calls)
-        8. Update RunResponse
-        9. Update Agent Memory
-        10. Calculate session metrics
-        11. Save session to storage
-        12. Save output to file if save_response_to_file is set
+        1. Reason about the task if reasoning is enabled
+        2. Generate a response from the Model (includes running function calls)
+        3. Update Agent Memory
+        4. Calculate session metrics
+        5. Save session to storage
+        6. Save output to file if save_response_to_file is set
         """
-
-        # 1. Prepare the Agent for the run
         if isinstance(self.memory, AgentMemory):
             self.memory = cast(AgentMemory, self.memory)
         else:
             self.memory = cast(Memory, self.memory)
-        # 1.2 Set streaming and stream intermediate steps
+        self.model = cast(Model, self.model)
 
-        self.stream = self.stream or (stream and self.is_streamable)
-        self.stream_intermediate_steps = self.stream_intermediate_steps or (stream_intermediate_steps and self.stream)
-        # 1.3 Create a run_id and RunResponse
-        self.run_id = str(uuid4())
+        # 1. Reason about the task if reasoning is enabled
+        if self.reasoning or self.reasoning_model is not None:
+            reasoning_generator = self.reason(run_messages=run_messages, session_id=session_id)
 
-        log_debug(f"Agent Run Start: {run_response.run_id}", center=True)
+            # Consume the generator without yielding
+            deque(reasoning_generator, maxlen=0)
 
-        # 2.1 Prepare arguments for the model
-        self.set_default_model()
-        response_format = self._get_response_format()
-        self.model = cast(Model, self.model)
+        # Get the index of the last "user" message in messages_for_run
+        # We track this so we can add messages after this index to the RunResponse and Memory
+        index_of_last_user_message = len(run_messages.messages)
 
-        self.determine_tools_for_model(
-            model=self.model,
-            session_id=session_id,
-            user_id=user_id,
-            async_mode=False,
-            knowledge_filters=knowledge_filters,
+        # 2. Generate a response from the Model (includes running function calls)
+        model_response: ModelResponse
+
+        # Get the model response
+        model_response = self.model.response(
+            messages=run_messages.messages,
+            response_format=response_format,
+            tools=self._tools_for_model,
+            functions=self._functions_for_model,
+            tool_choice=self.tool_choice,
+            tool_call_limit=self.tool_call_limit,
         )
-        run_response.model = self.model.id if self.model is not None else None
+        # Format tool calls if they exist
+        if model_response.tool_calls:
+            run_response.formatted_tool_calls = format_tool_calls(model_response.tool_calls)
```

Comment:
Not for now but should we try to keep all tool logic in the model? (looking at `format_tool_calls`)